### PR TITLE
[WIP] Aymeric's Stoch

### PR DESCRIPTION
**Pull request by DLWoodruff**
stoch admm

### CODE REVIEW

**Comment by DLWoodruff** [comment]
All tests pass on my computer. I think the test that is failing is due to some issues with the xpress solver or something.

File: examples/distr/distr_admm_cylinders.py [hunk 1]
```diff
@@ -27,6 +26,7 @@ def _parse_args():
     cfg.xhatxbar_args()
     cfg.lagrangian_args()
     cfg.ph_ob_args()
+    cfg.wxbar_read_write_args()
```
Comment (jeanpaulwatson) [inline examples/distr/distr_admm_cylinders.py:29]: Why is this here?

File: examples/stoch_distr/stoch_distr_ef.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+def _parse_args():
+    # create a config object and parse
+    cfg = config.Config()
+    stoch_distr.inparser_adder(cfg)
+    cfg.add_to_config("solver_name",
+                         description="Choice of the solver",
+                         domain=str,
+                         default=None)
+
+    cfg.parse_command_line("distr_ef2")
```
Comment (jeanpaulwatson) [inline examples/stoch_distr/stoch_distr_ef.py:31]: The "2" suffix is weird.

File: mpisppy/utils/stoch_admmWrapper.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        error_list1 = []
+        error_list2 = []
+        for sname,s in self.local_admm_stoch_subproblem_scenarios.items():
+            if verbose:
+                print(f"stoch_admmWrapper.assign_variable_probs is processing scenario: {sname}")
+            admm_subproblem_name = self.split_admm_stoch_subproblem_scenario_name(sname)[0]
+            # varlist[stage] will contain the variables at each stage
+            depth = len(s._mpisppy_node_list)+1
+            varlist = [[] for _ in range(depth)]
+            if s._mpisppy_probability == "uniform": #if there is a two-stage scenario defined by sputils.attach_root_node
```
Comment (jeanpaulwatson) [inline mpisppy/utils/stoch_admmWrapper.py:137]: Should be an assert for not having an _mpisppy_probability.

File: mpisppy/utils/stoch_admmWrapper.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+                        v.fix(0)
+                        self.varprob_dict[s].append((id(v),0))
+                    else:
+                        error_list2.append((sname,vstr))
+                varlist[stage-1].append(v)
+
+            # Create the new scenario tree node for admm_consensus
+            assert hasattr(s,"_mpisppy_node_list"), f"the scenario {sname} doesn't have any _mpisppy_node_list attribute"
+            parent = s._mpisppy_node_list[-1]
+            objfunc = pyo.Expression(expr=0) # The cost is spread on the branches which are the subproblems
```
Comment (jeanpaulwatson) [inline mpisppy/utils/stoch_admmWrapper.py:181]: Not sure where objfunc is used below - it is only passed in as an argument below.

File: mpisppy/utils/stoch_admmWrapper.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            else:
+                node_num = num_scen
+            node_name = parent.name + '_' + str(node_num) 
+
+            #could be more efficient with a dictionary rather than index
+            s._mpisppy_node_list.append(scenario_tree.ScenarioNode(
+                node_name,
+                1/self.number_admm_subproblems, #branching probability at this node
+                parent.stage + 1, # The stage is the stage of the previous leaves
+                objfunc * self.number_admm_subproblems,
```
Comment (jeanpaulwatson) [inline mpisppy/utils/stoch_admmWrapper.py:195]: Why multiply by number of subproblems?

**Review by jeanpaulwatson — APPROVED** [review]
